File: AkkaSupervisorExample/src/main/scala/org/akka/essentials/supervisor/example3/SupervisorActor.scala
package org.akka.essentials.supervisor.example3
import akka.actor.Actor
import akka.actor.ActorLogging
import akka.actor.Props
import akka.dispatch.Future
import akka.pattern.ask
import akka.util.Timeout
import akka.dispatch.Await

class SupervisorActor extends Actor with ActorLogging {
  import akka.actor.OneForOneStrategy
  import akka.actor.SupervisorStrategy._
  import akka.util.duration._
  import org.akka.essentials.supervisor.example3._

  var childActor = context.actorOf(Props[WorkerActor], name = "workerActor")
  val monitor = context.system.actorOf(Props[MonitorActor], name = "monitor")

  override def preStart() {
    monitor ! new RegisterWorker(childActor, self)
  }

  override val supervisorStrategy = OneForOneStrategy(maxNrOfRetries = 10, withinTimeRange = 10 seconds) {

    case _: ArithmeticException => Resume
    case _: NullPointerException => Restart
    case _: IllegalArgumentException => Stop
    case _: Exception => Escalate
  }

  def receive = {
    case result: Result =>
      childActor.tell(result, sender)
    case mesg: DeadWorker =>
      log.info("Got a DeadWorker message, restarting the worker")
      childActor = context.actorOf(Props[WorkerActor], name = "workerActor")
    case msg: Object =>
      childActor ! msg
  }
}

File: AkkaSupervisorExample/src/main/java/org/akka/essentials/supervisor/example3/SupervisorActor.java
package org.akka.essentials.supervisor.example3;

import static akka.actor.SupervisorStrategy.escalate;
import static akka.actor.SupervisorStrategy.restart;
import static akka.actor.SupervisorStrategy.resume;
import static akka.actor.SupervisorStrategy.stop;
import akka.actor.ActorRef;
import akka.actor.OneForOneStrategy;
import akka.actor.Props;
import akka.actor.SupervisorStrategy;
import akka.actor.SupervisorStrategy.Directive;
import akka.actor.UntypedActor;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.japi.Function;
import akka.util.Duration;

public class SupervisorActor extends UntypedActor {
	private LoggingAdapter log = Logging.getLogger(getContext().system(), this);
	public ActorRef workerActor = getContext().actorOf(
			new Props(WorkerActor.class), "workerActor");

	ActorRef monitor = getContext().system().actorOf(
			new Props(MonitorActor.class), "monitorActor");

	@Override
	public void preStart() {
		//MyActorSystem.monitor.tell(new RegisterWorker(workerActor, self()));
		monitor.tell(new RegisterWorker(workerActor, self()));
	}

	private static SupervisorStrategy strategy = new OneForOneStrategy(10,
			Duration.parse("10 second"), new Function<Throwable, Directive>() {
				public Directive apply(Throwable t) {
					if (t instanceof ArithmeticException) {
						return resume();
					} else if (t instanceof NullPointerException) {
						return restart();
					} else if (t instanceof IllegalArgumentException) {
						return stop();
					} else {
						return escalate();
					}
				}
			});

	@Override
	public SupervisorStrategy supervisorStrategy() {
		return strategy;
	}

	public void onReceive(Object o) throws Exception {
		if (o instanceof Result) {
			workerActor.tell(o, getSender());
		} else if (o instanceof DeadWorker) {
			log.info("Got a DeadWorker message, restarting the worker");
			workerActor = getContext().actorOf(new Props(WorkerActor.class),
					"workerActor");
		} else
			workerActor.tell(o);
	}

	public ActorRef getWorker(){
		return workerActor;
	}
}
